src/dune_engine/scheduler.ml

```diff
@@ -803,6 +865,9 @@ end = struct
     t.handler t.config Tick;
     match Event.Queue.next t.events with
     | Job_completed (job, proc_info) -> Fiber.Fill (job.ivar, proc_info)
+    | File_watcher_task job ->
+      job ();
```

Comment:
I'm trying to understand how `async_inotify` communicates with the scheduler.

IIUC, when `async_inotify` receive some data from inotify, it sends a `(unit -> unit)` job to the scheduler via `File_watcher_task`. This job:

- processes the data received from inotify and produce a list of `Async_inotify.Event.t`
- sends these events one by one to the scheduler via the `emit_event` callback
- this callback, provided by `dune_file_watcher`, packs the event into a list of a single element and adds this list to the event queue

So zooming out, when we receive an inotify event:

- async_inotify sends a `(unit -> unit)` callback to the scheduler
- the scheduler picks it up, execute it and this callback enqueues more things to the event queue
- the scheduler picks up the newly added events from the queue and process them

It feels like we could simplify this by making `async_inotify` send a `unit -> Event.t list` callback to the scheduler instead. WDYT?

Reply:
That is correct.
I agree we should be able to emit a batch of events instead of doing it one by one.
Looking at that...

Reply:
Pushed that change. Looks better. I also split `pump_events` into two parts (based on which thread they run in) and added an .mli (I forgot to bring that in from jane).

Reply:
That looks better indeed. Could we go one step further and send a `(unit -> Event.t list)` to the scheduler rather that send a `(unit -> unit)` that enqueues more events? That would remove one hoop. We might also be able to move the `process_inotify_event` to `Dune_file_watcher`, so that the scheduler never sees the low-level inotify events.

Reply:
Sounds good, trying that...

Reply:
@jeremiedimino, I pushed a commit (ba509cb2c96844dcd9c3ef9d1392751ef58aa39f) that makes this change and moves event processing code (including `ignored_files`) from the scheduler to file watcher. I like the direction this is going. In the current state there's still two separate functions in event queue for:

```
    val send_file_watcher_task : t -> (unit -> Dune_file_watcher.Event.t list) -> unit
    val send_file_watcher_events : t -> Dune_file_watcher.Event.t list -> unit
```

but only the former is exposed to the file watcher and we could probably get rid of it, but I thought I'd stop somewhere.

Reply:
Sounds good